File: app/src/main/java/net/tlalka/android/fiszki/domain/controllers/ListController.java
package net.tlalka.android.fiszki.domain.controllers;

import java.util.List;
import java.util.Map;
import javax.inject.Inject;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import net.tlalka.android.fiszki.core.annotations.ActivityScope;
import net.tlalka.android.fiszki.domain.services.LessonService;
import net.tlalka.android.fiszki.domain.services.StorageService;
import net.tlalka.android.fiszki.domain.services.WordService;
import net.tlalka.android.fiszki.model.entities.Lesson;
import net.tlalka.android.fiszki.model.entities.Word;

@ActivityScope
public class ListController {

    @Inject
    protected StorageService storageService;

    @Inject
    protected LessonService lessonService;

    @Inject
    protected WordService wordService;

    @Inject
    public ListController() {
    }

    public List<Lesson> getLessonList() {
        return lessonService.getLessons(storageService.getLanguage());
    }

    public Map<Long, List<Word>> getWordsMap() {
        return Stream.of(wordService.getWords(storageService.getLanguage()))
                .collect(Collectors.groupingBy(w -> w.getLesson().getId()));
    }

    public Map<Long, Word> getTranslationMap() {
        return Stream.of(wordService.getWords(storageService.getTranslation()))
                .collect(Collectors.toMap(w -> w.getCluster().getId()));
    }
}

File: app/src/main/java/net/tlalka/android/fiszki/domain/controllers/ListController.kt
package net.tlalka.android.fiszki.domain.controllers

import com.annimon.stream.Collectors
import com.annimon.stream.Stream
import net.tlalka.android.fiszki.core.annotations.ActivityScope
import net.tlalka.android.fiszki.domain.services.LessonService
import net.tlalka.android.fiszki.domain.services.StorageService
import net.tlalka.android.fiszki.domain.services.WordService
import net.tlalka.android.fiszki.model.entities.Lesson
import net.tlalka.android.fiszki.model.entities.Word
import javax.inject.Inject

@ActivityScope
class ListController @Inject constructor() {

    @Inject
    lateinit var storageService: StorageService

    @Inject
    lateinit var lessonService: LessonService

    @Inject
    lateinit var wordService: WordService

    fun getLessonList(): List<Lesson> {
        return lessonService.getLessons(storageService.language)
    }

    fun getWordsMap(): Map<Long, List<Word>> {
        return Stream
                .of(wordService.getWords(storageService.language))
                .collect(Collectors.groupingBy { w -> w.lesson.id })
    }

    fun getTranslationMap(): Map<Long, Word> {
        return Stream
                .of(wordService.getWords(storageService.translation))
                .collect(Collectors.toMap { w -> w.cluster.id })
    }
}
